File: domusviewer/app/src/main/java/it/achdjian/paolo/domusviewer/temperature/TemperatureLabel.java
package it.achdjian.paolo.domusviewer.temperature;

import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.support.annotation.NonNull;

import org.rajawali3d.materials.Material;
import org.rajawali3d.materials.textures.ATexture;
import org.rajawali3d.materials.textures.Texture;
import org.rajawali3d.math.vector.Vector3;
import org.rajawali3d.primitives.Plane;

import java.util.Locale;

/**
 * Created by Paolo Achdjian on 17/04/17.
 */

public class TemperatureLabel extends Plane {
    private float previousTemp = Float.NaN;
    public static float WIDTH =2f;
    public static float HEIGHT =0.5f;

    private static float getMaxWidth(RoomObject room){
        float minWidth = (float) (room.getMax().x-room.getMin().x);
        return minWidth < WIDTH ? minWidth : WIDTH;
    }

    private static float getMaxHeight(RoomObject room){
        float minHeight = (float) (room.getMax().y-room.getMin().y);
        return minHeight < HEIGHT ? minHeight : HEIGHT;
    }

    public TemperatureLabel(@NonNull RoomObject room) {
        super(getMaxWidth(room), getMaxHeight(room), 3, 3);
        String name = "TempSensor"+ room.name;
        setName(name);
        setText("N.D.", Color.GREEN);
        Vector3 pos = new Vector3(room.mean.x, room.mean.y, 0);
        setPosition(pos);
        setDoubleSided(true);
    }

    public void setTemperature(float temp, int background) {
        if (temp != previousTemp) {
            previousTemp = temp;
            String strTemp = String.format(Locale.US, "%.2f", temp);
            setText(strTemp, background);
        }
    }

    private void setText(@NonNull String text, int background) {
        Material temperatureMaterial = new Material(true);
        temperatureMaterial.setColor(background);
        Bitmap bitmap = textAsBitmap(text, Color.WHITE);
        try {
            int maxTextureName = 18;
            String name = getName().substring(0, getName().length() > maxTextureName ? maxTextureName : getName().length() );
            temperatureMaterial.addTexture(new Texture(name, bitmap));
            setMaterial(temperatureMaterial);
        } catch (ATexture.TextureException e) {
            e.printStackTrace();
        }
    }

    private Bitmap textAsBitmap(@NonNull String text, int textColor) {
        Paint paint = new Paint();
        paint.setTextSize(30);
        paint.setStrokeWidth(10);
        paint.setColor(textColor);
        float baseline = (int) (-paint.ascent() + 0.5f);


        Bitmap image = Bitmap.createBitmap(128, 32, Bitmap.Config.ARGB_8888);
        Canvas canvas = new Canvas(image);
        int widthText = (int) paint.measureText(text);
        canvas.drawText(text, (128 - widthText) / 2, baseline, paint);
        return image;
    }


}

File: temperature_monitor/app/src/main/java/it/achdjian/paolo/temperaturemonitor/rajawali/TemperatureLabel.kt
package it.achdjian.paolo.temperaturemonitor.rajawali

import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import org.rajawali3d.materials.Material
import org.rajawali3d.materials.textures.ATexture
import org.rajawali3d.materials.textures.Texture
import org.rajawali3d.math.vector.Vector3
import org.rajawali3d.primitives.Plane
import java.util.*

/**
 * Created by Paolo Achdjian on 17/04/17.
 */

class TemperatureLabel(room: RoomObject) : Plane(getMaxWidth(room), getMaxHeight(room), 3, 3) {
    private var previousTemp = java.lang.Float.NaN

    val pos: Vector3
        get() = position

    init {
        val name = "TempSensor" + room.name
        setName(name)
        setText("N.D.", Color.GREEN)
        val pos = Vector3(room.mean.x, room.mean.y, room.max.z)
        position = pos
        isDoubleSided = true
    }

    fun setTemperature(temp: Float, background: Int) {
        if (temp != previousTemp) {
            previousTemp = temp
            val strTemp = String.format(Locale.US, "%.2f", temp)
            setText(strTemp, background)
        }
    }

    private fun setText(text: String, background: Int) {
        val temperatureMaterial = Material(true)
        temperatureMaterial.color = background
        val bitmap = textAsBitmap(text, Color.WHITE)
        try {
            val maxTextureName = 18
            val name = name.substring(0, if (name.length > maxTextureName) maxTextureName else name.length)
            temperatureMaterial.addTexture(Texture(name, bitmap))
            material = temperatureMaterial
        } catch (e: ATexture.TextureException) {
            e.printStackTrace()
        }

    }

    private fun textAsBitmap(text: String, textColor: Int): Bitmap {
        val paint = Paint()
        paint.textSize = 30f
        paint.strokeWidth = 10f
        paint.color = textColor
        val baseline = (-paint.ascent() + 0.5f).toInt().toFloat()


        val image = Bitmap.createBitmap(128, 32, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(image)
        val widthText = paint.measureText(text).toInt()
        canvas.drawText(text, ((128 - widthText) / 2).toFloat(), baseline, paint)
        return image
    }

    companion object {
        var WIDTH = 1.5f
        var HEIGHT = 0.4f

        private fun getMaxWidth(room: RoomObject): Float {
            val minWidth = (room.max.x - room.min.x).toFloat()
            return if (minWidth < WIDTH) minWidth else WIDTH
        }

        private fun getMaxHeight(room: RoomObject): Float {
            val minHeight = (room.max.y - room.min.y).toFloat()
            return if (minHeight < HEIGHT) minHeight else HEIGHT
        }
    }


}
